File: src/main/java/com/rimlang/rim/syntax/SyntaxAnalyzer.java
package com.rimlang.rim.syntax;

import com.rimlang.rim.lexer.StringToken;
import com.rimlang.rim.lexer.Token;
import com.rimlang.rim.lexer.TokenType;
import com.rimlang.rim.util.Strings;
import com.rimlang.rim.util.TokenUtil;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.stream.Collectors;

public class SyntaxAnalyzer {

    public static List<SyntaxNode> analyze(List<Token> code) throws RimSyntaxException {
        List<SyntaxNode> result = new ArrayList<>();
        List<Token> codeLine = new ArrayList<>();
        int beginningOfLine = 0;
        for (int i = 0; i < code.size(); i++) {
            Token t = code.get(i);
            if (t.getType() != TokenType.EOL) {
                codeLine.add(t);
            }
            if (t.getType() == TokenType.EOL || i == code.size() - 1) {
                if (codeLine.isEmpty()) continue;
                result.add(processLine(codeLine, code, codeLine.get(0).getLine(), beginningOfLine));
                codeLine.clear();
                beginningOfLine = i + 1;
            }
        }
        return result;
    }

    private static @NotNull SyntaxNode processLine(List<Token> codeLine, List<Token> code, int line, int beginningOfLine) throws RimSyntaxException {
        TokenType first = codeLine.get(0).getType();
        switch(first) {
            case FUNCTION, COMMAND -> {
                String lineType = Strings.title(first.name());
                if (codeLine.get(1).getType() != TokenType.ID) {
                    throw new RimSyntaxException(lineType + " definition must have a name", line);
                }
                StringToken name = (StringToken) codeLine.get(1);
                if (codeLine.get(2).getType() != TokenType.PARENTHESES) {
                    throw new RimSyntaxException(lineType + " definition must have parentheses, even if there are zero arguments", line);
                }
                ArgsExpression args = new ArgsExpression(TokenUtil.getGroup(codeLine.get(2)));
                Token body = codeLine.get(codeLine.size() - 1);
                if (body.getType() != TokenType.BRACE) {
                    throw new RimSyntaxException(lineType + " must have a code body", line);
                }
                List<SyntaxNode> syntaxNodes = analyze(TokenUtil.getGroup(body));
                if (first == TokenType.COMMAND) { // Command doesn't have a return type
                    return new CommandDefinitionSyntaxNode(name, args, syntaxNodes);
                }

                StringToken returnType = null;
                if (codeLine.get(3).getType() == TokenType.COLON) {
                    if (codeLine.get(4).getType() != TokenType.TYPE) {
                        throw new RimSyntaxException(lineType + " has a colon, but no return type", line);
                    }
                    returnType = (StringToken) codeLine.get(4);
                }
                if (returnType == null) {
                    returnType = new StringToken(TokenType.TYPE, "void", line, -1);
                }
                return new FunctionDefinitionSyntaxNode(name, args, returnType, syntaxNodes);
            }
            case ID -> {
                TokenType second = codeLine.get(1).getType();
                if (second == TokenType.ASSIGNMENT_OPERATOR) {
                    return new VariableDefinitionSyntaxNode((StringToken) codeLine.get(0), new GenericExpression(TokenUtil.sub(codeLine, 2, codeLine.size())));
                } else if (second == TokenType.DOT) {
                    return new ExpressionSyntaxNode(codeLine);
                } else if (second == TokenType.PARENTHESES) {
                    List<Token> argTokens = TokenUtil.getGroup(codeLine.get(1));
                    assert argTokens != null;
                    List<GenericExpression> args = TokenUtil.split(argTokens, TokenType.COMMA)
                            .stream().map(GenericExpression::new).collect(Collectors.toList());
                    if (codeLine.size() <= 2 || codeLine.get(2).getType() != TokenType.BRACE) {
                        return new ExpressionSyntaxNode(codeLine);
                    }
                    List<Token> tokens = TokenUtil.getGroup(codeLine.get(2));
                    assert tokens != null;
                    if (codeLine.get(0) instanceof StringToken) {
                        StringToken stringToken = (StringToken) codeLine.get(0);
                        return new ComplexFunctionCallSyntaxNode(stringToken, args, analyze(tokens));
                    } else {
                        throw new RimSyntaxException("Function call must have the function name", line);
                    }
                }
            }
            case IF -> {
                Token body = codeLine.get(codeLine.size() - 1);
                if (body.getType() != TokenType.BRACE) {
                    throw new RimSyntaxException("If statement must have code body", line);
                }
                // TODO: Add else and elif

                return new IfSyntaxNode(
                        new BooleanExpression(TokenUtil.sub(codeLine, 1, codeLine.size() - 1)),
                        analyze(TokenUtil.getGroup(body))
                );
            }
            case WHILE -> {
                Token body = codeLine.get(codeLine.size() - 1);
                if (body.getType() != TokenType.BRACE) {
                    throw new RimSyntaxException("While statement must have code body", line);
                }
                return new WhileSyntaxNode(
                        new BooleanExpression(TokenUtil.sub(codeLine, 1, codeLine.size() - 1)),
                        analyze(TokenUtil.getGroup(body))
                );
            }
            case MATCH -> {
                Token bodyToken = codeLine.get(codeLine.size() - 1);
                if (bodyToken.getType() != TokenType.BRACE) {
                    throw new RimSyntaxException("Match statement must have code body", line);
                }
                List<Token> body = TokenUtil.getGroup(bodyToken);
                HashMap<List<Token>, List<SyntaxNode>> branches = new HashMap<>();
                TokenUtil.split(body, TokenType.EOL).forEach(tokens -> {
                    Token branchBodyToken = tokens.get(tokens.size() - 1);
                    try {
                        if (branchBodyToken.getType() != TokenType.BRACE) {
                            throw new RimSyntaxException("Match branch must have code body", branchBodyToken.getLine());
                        }
                        branches.put(
                                TokenUtil.sub(tokens, 0, tokens.size() - 1),
                                analyze(TokenUtil.getGroup(branchBodyToken))
                        );
                    } catch (RimSyntaxException e) {
                        throw new RuntimeException(e);
                    }
                });
                return new MatchSyntaxNode(
                        new GenericExpression(TokenUtil.sub(codeLine, 1, codeLine.size() - 1)),
                        branches
                );
            }
            case RETURN -> {
                return new ReturnSyntaxNode(new GenericExpression(TokenUtil.sub(code, beginningOfLine + 1)));
            }
            case FOR -> {
                Token body = codeLine.get(codeLine.size() - 1);
                if (body.getType() != TokenType.BRACE) {
                    throw new RimSyntaxException("For each statement must have code body", line);
                }
                if (codeLine.get(2).getType() != TokenType.IN) {
                    throw new RimSyntaxException("For loop must have 'in' token", line);
                }
                return new ForeachSyntaxNode(
                        codeLine.get(1),
                        new GenericExpression(TokenUtil.sub(codeLine, 3, codeLine.size() - 1)),
                        analyze(TokenUtil.getGroup(body))
                );
            }
        }
        System.out.println(codeLine);
        throw new RimSyntaxException("Illegal start to line", line);
    }

}

File: src/main/java/com/rimlang/rim/syntax/SyntaxAnalyzer.kt
package com.rimlang.rim.syntax

import com.rimlang.rim.lexer.StringToken
import com.rimlang.rim.lexer.Token
import com.rimlang.rim.lexer.TokenType
import com.rimlang.rim.util.*

@Throws(RimSyntaxException::class)
fun analyze(code: List<Token>): List<SyntaxNode> {
    val result = mutableListOf<SyntaxNode>()
    val codeLine = mutableListOf<Token>()
    var beginningOfLine = 0
    for (i in code.indices) {
        val t = code[i]
        if (t.type !== TokenType.EOL) {
            codeLine.add(t)
        }
        if (t.type === TokenType.EOL || i == code.size - 1) {
            if (codeLine.isEmpty()) continue
            result += processLine(codeLine, code, codeLine[0].line, beginningOfLine)
            codeLine.clear()
            beginningOfLine = i + 1
        }
    }
    return result
}

@Throws(RimSyntaxException::class)
private fun processLine(codeLine: List<Token>, code: List<Token>, line: Int, beginningOfLine: Int): SyntaxNode {
    when (val first = codeLine[0].type) {
        TokenType.FUNCTION, TokenType.COMMAND -> {
            val lineType = first.name.title()
            if (codeLine[1].type !== TokenType.ID) {
                throw RimSyntaxException("$lineType definition must have a name", line)
            }
            val name = codeLine[1] as StringToken
            if (codeLine[2].type !== TokenType.PARENTHESES) {
                throw RimSyntaxException(
                    "$lineType definition must have parentheses, even if there are zero arguments",
                    line
                )
            }
            val args = ArgsExpression(getGroup(codeLine[2]))
            val body = codeLine[codeLine.size - 1]
            if (body.type !== TokenType.BRACE) {
                throw RimSyntaxException("$lineType must have a code body", line)
            }
            val syntaxNodes = analyze(getGroup(body))
            if (first === TokenType.COMMAND) { // Command doesn't have a return type
                return CommandDefinitionSyntaxNode(name, args, syntaxNodes)
            }
            val returnType = if (codeLine[3].type === TokenType.COLON) {
                if (codeLine[4].type !== TokenType.TYPE) {
                    throw RimSyntaxException("$lineType has a colon, but no return type", line)
                }
                codeLine[4] as StringToken
            } else {
                StringToken(TokenType.TYPE, "void", line, -1)
            }
            return FunctionDefinitionSyntaxNode(name, args, returnType, syntaxNodes)
        }
        TokenType.ID -> {
            when (val second = codeLine[1].type) {
                TokenType.ASSIGNMENT_OPERATOR -> {
                    return VariableDefinitionSyntaxNode(
                        codeLine[0] as StringToken,
                        GenericExpression(sub(codeLine, 2, codeLine.size))
                    )
                }
                TokenType.DOT -> {
                    println(codeLine)
                    return ExpressionSyntaxNode(codeLine)
                }
                TokenType.PARENTHESES -> {
                    val argTokens = getGroup(codeLine[1])
                    val args = argTokens.split(TokenType.COMMA).map(::GenericExpression)
                    if (codeLine.size <= 2 || codeLine[2].type !== TokenType.BRACE) {
                        println(codeLine)
                        return ExpressionSyntaxNode(codeLine)
                    }
                    val tokens = getGroup(codeLine[2])
                    return if (codeLine[0] is StringToken) {
                        val stringToken = codeLine[0] as StringToken
                        ComplexFunctionCallSyntaxNode(stringToken, args, analyze(tokens))
                    } else {
                        throw RimSyntaxException("Function call must have the function name", line)
                    }
                }
                else -> {
                    throw RimSyntaxException("Unexpected token after identifier of type ${second.name}", line)
                }
            }
        }
        TokenType.IF -> {
            val branches = mutableListOf<ConditionalBranch>()
            val split = codeLine.split(TokenType.BRACE, true)
            for (i in split.indices) {
                val branch = split[i]
                if (branch.isEmpty()) continue
                if (branch[0].type !in listOf(TokenType.IF, TokenType.ELSE, TokenType.ELIF)) {
                    throw RimSyntaxException("If statement can't have a branch starting with ${branch[0].type.name.lowercase()}.", line)
                }
                if (i != split.lastIndex && branch[0].type == TokenType.ELSE) {
                    throw RimSyntaxException("Else statement can't be followed by another branch", line)
                }
                if (i != 0 && branch[0].type == TokenType.IF) {
                    throw RimSyntaxException("If statement can't be preceded by another branch", line)
                }
                branches += ConditionalBranch(
                    branch[0] as StringToken,
                    BooleanExpression(branch.drop(1).dropLast(1)),
                    analyze(getGroup(branch.last()))
                )
            }
            if (branches.isEmpty()) {
                throw RimSyntaxException("If statement must have a code body", line)
            }
            return IfSyntaxNode(branches)
        }
        TokenType.WHILE -> {
            val body = codeLine.last()
            if (body.type !== TokenType.BRACE) {
                throw RimSyntaxException("While statement must have code body", line)
            }
            return WhileSyntaxNode(
                BooleanExpression(sub(codeLine, 1, codeLine.size - 1)),
                analyze(getGroup(body))
            )
        }
        TokenType.MATCH -> {
            val bodyToken = codeLine.last()
            if (bodyToken.type !== TokenType.BRACE) {
                throw RimSyntaxException("Match statement must have code body", line)
            }
            val body = getGroup(bodyToken)
            val branches = HashMap<List<Token>, List<SyntaxNode>>()
            body.split(TokenType.EOL).forEach {
                val branchBodyToken = it.last()
                if (branchBodyToken.type !== TokenType.BRACE) {
                    throw RimSyntaxException("Match branch must have code body", branchBodyToken.line)
                }
                branches[sub(it, 0, it.lastIndex)] =
                    analyze(getGroup(branchBodyToken))
            }
            return MatchSyntaxNode(
                GenericExpression(sub(codeLine, 1, codeLine.size - 1)),
                branches
            )
        }
        TokenType.RETURN -> {
            return ReturnSyntaxNode(GenericExpression(sub(code, beginningOfLine + 2)))
        }
        TokenType.FOR -> {
            val body = codeLine[codeLine.size - 1]
            if (body.type !== TokenType.BRACE) {
                throw RimSyntaxException("For each statement must have code body", line)
            }
            if (codeLine[2].type !== TokenType.IN) {
                throw RimSyntaxException("For loop must have 'in' token", line)
            }
            return ForeachSyntaxNode(
                codeLine[1] as StringToken,
                GenericExpression(sub(codeLine, 3, codeLine.size - 1)),
                analyze(getGroup(body))
            )
        }
        else -> {
            println(codeLine)
            throw RimSyntaxException("Illegal start to line", line)
        }
    }
}
